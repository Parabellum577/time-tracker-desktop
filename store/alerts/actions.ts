import { createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { IAlert } from './types'

export const showAlertMessage = createAsyncAction(
  'notifications/SHOW_ALERT_MESSAGE_REQUEST',
  'notifications/SHOW_ALERT_MESSAGE_SUCCESS',
  'notifications/SHOW_ALERT_MESSAGE_FAILURE',
)<IAlert, IAlert, IError>()

export const clearAlertMessage = createAsyncAction(
  'notifications/CLEAR_ALERT_MESSAGE_REQUEST',
  'notifications/CLEAR_ALERT_MESSAGE_SUCCESS',
  'notifications/CLEAR_ALERT_MESSAGE_FAILURE',
)<void, void, IError>()
