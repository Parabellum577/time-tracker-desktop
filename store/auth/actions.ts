import { createAction, createAsyncAction } from 'typesafe-actions'
import { IError } from '@types'
import { ILoginUserPayload, ILoginOTSPayload } from '@services/api-types'

export const loginUser = createAsyncAction('auth/LOGIN_REQUEST', 'auth/LOGIN_SUCCESS', 'auth/LOGIN_FAILURE')<
  ILoginUserPayload,
  void,
  IError
>()

export const loginUserViaOAuth = createAsyncAction(
  'auth/LOGIN_VIA_OAUTH_REQUEST',
  'auth/LOGIN_VIA_OAUTH_SUCCESS',
  'auth/LOGIN_VIA_OAUTH_FAILURE',
)<void, void, IError>()

export const otsLogin = createAsyncAction('auth/OTS_LOGIN_REQUEST', 'auth/OTS_LOGIN_SUCCESS', 'auth/OTS_LOGIN_FAILURE')<
  ILoginOTSPayload,
  ILoginOTSPayload,
  IError
>()

export const loadingDataAfterLogin = createAction('auth/LOADING_DATA_AFTER_LOGIN', resolve => (isLoading: boolean) =>
  resolve(isLoading),
)

export const signOut = createAction('auth/SIGN_OUT', resolve => () => resolve())
