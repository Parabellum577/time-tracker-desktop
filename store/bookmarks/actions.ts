import { createAsyncAction, createAction } from 'typesafe-actions'
import { IError, IFullIssue } from '@types'

export const getBookmarks = createAsyncAction(
  'bookmarks/GET_BOOKMARKS_REQUEST',
  'bookmarks/GET_BOOKMARKS_SUCCESS',
  'bookmarks/GET_BOOKMARKS_FAILURE',
)<void, IFullIssue[], IError>()

export const handleBookmark = createAsyncAction(
  'bookmarks/HANDLE_BOOKMARK_REQUEST',
  'bookmarks/HANDLE_BOOKMARK_SUCCESS',
  'bookmarks/HANDLE_BOOKMARK_FAILURE',
)<IFullIssue, IFullIssue, IError>()

export const clearBookmarks = createAction('plannings/CLEAR_BOOKMARKS', action => () => action())
