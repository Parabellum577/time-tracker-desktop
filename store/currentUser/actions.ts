import { createAction, createAsyncAction } from 'typesafe-actions'
import { IUser, IError, IUpdateUser, ISettingsItem } from '@types'

export const getCurrentUser = createAsyncAction(
  'user/GET_CURRENT_USER_REQUEST',
  'user/GET_CURRENT_USER_SUCCESS',
  'user/GET_CURRENT_USER_FAILURE',
)<void, IUser, IError>()

export const updateCurrentUser = createAction('user/UPDATE_CURRENT_USER', action => (userData: IUpdateUser) =>
  action({
    Avatar: userData.Avatar,
    FirstName: userData.FirstName,
    LastName: userData.LastName,
    Phone: userData.Phone,
    Skype: userData.Skype,
    Role: userData.Role,
  }),
)

export const handleUserSettings = createAsyncAction(
  'user/HANDLE_USER_SETTINGS_REQUEST',
  'user/HANDLE_USER_SETTINGS_SUCCESS',
  'user/HANDLE_USER_SETTINGS_FAILURE',
)<ISettingsItem[], ISettingsItem[], IError>()

export const getUserSettings = createAsyncAction(
  'user/GET_USER_SETTINGS_REQUEST',
  'user/GET_USER_SETTINGS_SUCCESS',
  'user/GET_USER_SETTINGS_FAILURE',
)<void, void, IError>()

export const setDefaultUserSettings = createAction('user/SET_DEFAULT_USER_SETTINGS', action => action)

export const activateUserAccount = createAction('user/ACTIVATE_USER_ACCOUNT', action => action)
