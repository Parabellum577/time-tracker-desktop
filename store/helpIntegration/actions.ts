import { createAction } from 'typesafe-actions'
import { IHintType } from './types'

export const showHelpWindow = createAction('helpIntegration/SHOW_INSTRUCTIONS', action => (template: IHintType) =>
  action(template),
)

export const closeHelpWindow = createAction('helpIntegration/HIDE_INSTRUCTIONS', action => action)
