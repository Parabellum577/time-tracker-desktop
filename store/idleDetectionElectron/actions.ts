import { createAction } from 'typesafe-actions'
import { INotification } from '../notifications/types'

export const idleDetected = createAction('idle/IDLE_DETECTED', action => (idleTime: number) => action(idleTime))

export const activityDetected = createAction('idle/ACTIVITY_DETECTED', action => () => action())
