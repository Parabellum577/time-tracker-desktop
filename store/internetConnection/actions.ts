import { createAction } from 'typesafe-actions'

export const changeConnectionStatus = createAction(
  'internet_connection/CONNECTION_STATUS_CHANGED',
  action => (isOnline: boolean) => action({ isOnline }),
)
