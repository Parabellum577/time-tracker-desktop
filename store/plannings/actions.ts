import {
  ICreatePlanningPayload,
  ISetExtraPayload,
  IStopPlanning,
  ISetStatusPayload,
  ISetPriorityPayload,
} from '@api-types'
import { createAction, createAsyncAction } from 'typesafe-actions'
import { IPlanning, IError, IReportColdPlannings, IReportManually, ITimeSummary } from '@types'

export const getOpenPlannings = createAsyncAction(
  'plannings/GET_OPEN_PLANNINGS_REQUEST',
  'plannings/GET_OPEN_PLANNINGS_SUCCESS',
  'plannings/GET_OPEN_PLANNINGS_FAILURE',
)<void, IPlanning[], IError>()

export const getColdPlannings = createAsyncAction(
  'plannings/GET_COLD_PLANNINGS_REQUEST',
  'plannings/GET_COLD_PLANNINGS_SUCCESS',
  'plannings/GET_COLD_PLANNINGS_FAILURE',
)<void, IPlanning[], IError>()

export const getClosedPlannings = createAsyncAction(
  'plannings/GET_CLOSED_PLANNINGS_REQUEST',
  'plannings/GET_CLOSED_PLANNINGS_SUCCESS',
  'plannings/GET_CLOSED_PLANNINGS_FAILURE',
)<void, IPlanning[], IError>()

export const startPlanning = createAsyncAction(
  'plannings/START_PLANNING_REQUEST',
  'plannings/START_PLANNING_SUCCESS',
  'plannings/START_PLANNING_FAILURE',
)<{ PlanningID: number; TracingID?: string }, number, IError>()

export const startPlanningSync = createAction('plannings/START_PLANNING_SYNC', action => (PlanningID: number) =>
  action(PlanningID),
)

export const stopPlanning = createAsyncAction(
  'plannings/STOP_PLANNING_REQUEST',
  'plannings/STOP_PLANNING_SUCCESS',
  'plannings/STOP_PLANNING_FAILURE',
)<IStopPlanning | void, void, IError>()

export const startColdPlanning = createAsyncAction(
  'plannings/START_COLD_PLANNING_REQUEST',
  'plannings/START_COLD_PLANNING_SUCCESS',
  'plannings/START_COLD_PLANNING_FAILURE',
)<number | void, number, IError>()

export const stopColdPlanning = createAsyncAction(
  'plannings/STOP_COLD_PLANNING_REQUEST',
  'plannings/STOP_COLD_PLANNING_SUCCESS',
  'plannings/STOP_COLD_PLANNING_FAILURE',
)<void, void, IError>()

export const createPlannings = createAsyncAction(
  'plannings/CREATE_PLANNING_REQUEST',
  'plannings/CREATE_PLANNING_SUCCESS',
  'plannings/CREATE_PLANNING_FAILURE',
)<Array<{ task: ICreatePlanningPayload; isStarted: boolean }>, void, IError>()

export const timeTickInActivePlanning = createAction(
  'plannings/TIME_TICK_IN_ACTIVE_PLANNING',
  action => (PlanningID: number, field: 'SpentOnline' | 'SpentOffline', additionalTime: number) =>
    action({ PlanningID, field, additionalTime }),
)

export const setSpentTime = createAction(
  'plannings/SET_SPENT_TIME',
  action => (PlanningID: number, SpentTime: number) => action({ PlanningID, SpentTime }),
)

export const setExtra = createAction('plannings/SET_EXTRA_PLANNING', action => (Planning: ISetExtraPayload) =>
  action({
    Estimation: Planning.Estimation,
    PlanningID: Planning.PlanningID,
    Reason: Planning.Reason,
  }),
)

export const coldPlanningRemap = createAction('plannings/COLD_PLANNING_REMAP', action => (From: number, To: number) =>
  action({ From, To }),
)

export const getTimeSummary = createAsyncAction(
  'plannings/TIME_SUMMARY_REQUEST',
  'plannings/TIME_SUMMARY_SUCCESS',
  'plannings/TIME_SUMMARY_FAILURE',
)<null, ITimeSummary, IError>()

export const closePlanning = createAsyncAction(
  'plannings/CLOSE_PLANNING_REQUEST',
  'plannings/CLOSE_PLANNING_SUCCESS',
  'plannings/CLOSE_PLANNING_FAILURE',
)<IPlanning, IPlanning, IError>()

export const addHighlightedPlanning = createAction('plannings/ADD_HIGHLIGHTED_PLANNING', action => (ID: string) =>
  action(ID),
)

export const removeHighlightedPlanning = createAction('plannings/REMOVE_HIGHLIGHTED_PLANNING', action => (ID: string) =>
  action(ID),
)

export const planningExpired = createAction('plannings/PLANNING_EXPIRED', action => (planning: IPlanning) =>
  action(planning),
)

export const closeColdPlanning = createAction('plannings/CHANGE_COLD_PLANNING_STATUS', action => (planningID: number) =>
  action(planningID),
)

export const reportColdPlannings = createAsyncAction(
  'plannings/REPORT_COLD_PLANNINGS_REQUEST',
  'plannings/REPORT_COLD_PLANNINGS_SUCCESS',
  'plannings/REPORT_COLD_PLANNINGS_FAILURE',
)<IReportColdPlannings, void, IError>()

export const reportManually = createAsyncAction(
  'plannings/REPORT_REPORT_MANUALLY_REQUEST',
  'plannings/REPORT_REPORT_MANUALLY_SUCCESS',
  'plannings/REPORT_REPORT_MANUALLY_FAILURE',
)<IReportManually, void, IError>()

export const clearPlannings = createAction('plannings/CLEAR_PLANNINGS', action => () => action())

export const setStatus = createAction('plannings/SET_STATUS', action => (payload: ISetStatusPayload) => action(payload))

export const setPriority = createAction('plannings/SET_PRIORITY', action => (payload: ISetPriorityPayload) =>
  action(payload),
)
