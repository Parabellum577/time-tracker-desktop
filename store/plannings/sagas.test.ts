import { take, select, resetMocks, fork, cancel, call, put, delay, cancelled } from '@test-helpers/effects.mock'
import { SIMPLE_PLANNING } from '@test-helpers/sample'
import { setResponse } from '@test-helpers/api.mock'
import time from '@test-helpers/time.mock'

import {
  getOpenPlanningsSaga,
  getOpenPlannings,
  startStopPlanningSaga,
  timeTickInActivePlanningLoop,
  createPlanning,
} from './sagas'
import * as actions from './actions'

describe('[Store] Plannings Sagas', () => {
  beforeEach(() => {
    resetMocks()
  })

  xit('[Success] Should handle GetOpenPlannings Request', () => {
    const generator = getOpenPlanningsSaga()
    generator.next()
    expect(call).toBeCalledTimes(1)
    expect(fork).toBeCalledTimes(1)
    generator.next()
    expect(take).toBeCalledTimes(4)
    generator.next()
    expect(generator.next().done).toBe(true)
  })

  it('[Success] Should emit GetOpenPlannings', () => {
    const openPlannings = [
      {
        ...SIMPLE_PLANNING,
        ID: 1,
      },
      {
        ...SIMPLE_PLANNING,
        ID: 2,
      },
    ]
    const callback = jest.fn(() => openPlannings)
    setResponse('planning', 'GetOpenPlannings', callback)
    const generator = getOpenPlannings()
    generator.next()
    expect(call).toBeCalledTimes(1)
    expect(call).toBeCalledWith(callback)
    generator.next(openPlannings as any)
    expect(put).toBeCalledTimes(1)
    expect(generator.next().done).toBe(true)
  })

  it('[Failed] Should emit GetOpenPlannings', () => {
    call.mockImplementation(() => {
      throw {
        code: 100,
        message: 'SOME_ERROR',
      }
    })
    put.mockImplementation(action => action)
    const generator = getOpenPlannings()
    expect(generator.next()).toMatchSnapshot()
    expect(generator.next().done).toBe(true)
  })

  xit('[Success] Should handle Start/Stop Planning Request', () => {
    // start planning
    let syncSpentTimeWithBackEndTask = jest.fn()
    let timeTrackingTask = jest.fn()
    let openPlannings = [
      {
        ...SIMPLE_PLANNING,
        ID: 1,
      },
      {
        ...SIMPLE_PLANNING,
        ID: 2,
      },
    ]
    const generator = startStopPlanningSaga()
    generator.next()
    expect(take).toBeCalledTimes(2)
    generator.next({
      startPlanningAction: {
        payload: 2,
      },
    })
    generator.next(openPlannings)
    generator.next([])
    expect(select).toBeCalledTimes(2)
    expect(call).toBeCalledTimes(1)
    generator.next()
    expect(fork).toBeCalledTimes(1)
    generator.next(syncSpentTimeWithBackEndTask)
    expect(fork).toBeCalledTimes(2)
    // stop planning
    generator.next(timeTrackingTask)
    generator.next({ startPlanningAction: false })
    openPlannings = [
      {
        ...SIMPLE_PLANNING,
        ID: 1,
      },
      {
        ...SIMPLE_PLANNING,
        Active: true,
        ID: 2,
      },
    ]
    generator.next(openPlannings)
    generator.next([])
    expect(take).toBeCalledTimes(4)
    expect(cancel).toBeCalledTimes(1)
    generator.next()
    expect(cancel).toBeCalledTimes(2)
    generator.next()
    expect(call).toBeCalledTimes(2)
    // start planning
    syncSpentTimeWithBackEndTask = jest.fn()
    timeTrackingTask = jest.fn()
    openPlannings = [
      {
        ...SIMPLE_PLANNING,
        ID: 1,
      },
      {
        ...SIMPLE_PLANNING,
        ID: 2,
      },
    ]
    generator.next()
    generator.next({
      startPlanningAction: {
        payload: 2,
      },
    })
    generator.next(openPlannings)
    generator.next([])
    generator.next([])
    generator.next()
    generator.next()
    generator.next(syncSpentTimeWithBackEndTask)
    generator.next(timeTrackingTask)
    // start planning when hasActivePlanning
    syncSpentTimeWithBackEndTask = jest.fn()
    timeTrackingTask = jest.fn()

    openPlannings = [
      {
        ...SIMPLE_PLANNING,
        ID: 1,
      },
      {
        ...SIMPLE_PLANNING,
        Active: true,
        ID: 2,
      },
    ]
    generator.next({
      startPlanningAction: {
        payload: 1,
      },
    })
    generator.next(openPlannings)
    generator.next([])
    expect(cancel).toBeCalledTimes(5)
    generator.next()
    expect(cancel).toBeCalledTimes(6)
    expect(call).toBeCalledTimes(3)
    generator.next()
    expect(call).toBeCalledTimes(4)
    generator.next()
    expect(call).toBeCalledTimes(5)
    generator.next()
    expect(fork).toBeCalledTimes(5)
    generator.next(syncSpentTimeWithBackEndTask)
    expect(fork).toBeCalledTimes(6)
  })

  // it('[Success] Should emit Start Planning', () => {
  //   const callback = jest.fn()
  //   time.now.mockReturnValue(1550135924)
  //   setResponse('planning', 'SetActive', callback)
  //   const generator = startPlanning(2)
  //   generator.next()
  //   expect(call).toBeCalledTimes(1)
  //   expect(call).toBeCalledWith(callback, { PlanningID: 2, Time: time.now() })
  //   generator.next()
  //   expect(put).toBeCalledTimes(1)
  //   expect(generator.next().done).toBe(true)
  // })

  // it('[Failed] Should emit Start Planning', () => {
  //   call.mockImplementation(() => {
  //     throw {
  //       code: 100,
  //       message: 'SOME_ERROR',
  //     }
  //   })
  //   time.now.mockReturnValueOnce(1550135924)
  //   const generator = startPlanning(1)
  //   generator.next()
  //   generator.next()
  //   expect(put).toBeCalledTimes(1)
  //   expect(generator.next().done).toBe(true)
  // })

  // it('[Success] Should emit Stop Planning', () => {
  //   const callback = jest.fn()
  //   setResponse('planning', 'SetActive', callback)
  //   time.now.mockReturnValue(1550135924)
  //   const generator = stopPlanning(1)
  //   generator.next()
  //   expect(call).toBeCalledTimes(1)
  //   expect(call).toBeCalledWith(callback, { PlanningID: 0, Time: time.now() })
  //   generator.next()
  //   expect(put).toBeCalledTimes(1)
  //   expect(generator.next().done).toBe(true)
  // })

  xit('[Success] Should handle createPlanning', () => {
    const payload: any = { task: {}, isStarted: false }
    let generator = createPlanning(payload)
    generator.next()
    expect(call).toBeCalled()
    generator.next()
    expect(put).not.toBeCalled()
    expect(generator.next().done).toBe(true)

    generator = createPlanning({ ...payload, isStarted: true })
    generator.next()
    expect(call).toBeCalled()
    generator.next()
    generator.next()
    expect(put).toBeCalled()
    expect(generator.next().done).toBe(true)
  })

  xit('[Success] Should update spent time every seconds', () => {
    const callback = jest.fn()
    setResponse('planning', 'SetSpent', callback)
    let timeNow = 1550135924000
    time.nowExact.mockReturnValue(timeNow)
    const generator = timeTickInActivePlanningLoop(2)
    generator.next()
    let lastArgs = put.mock.calls[put.mock.calls.length - 1]
    expect(lastArgs).toMatchSnapshot()
    generator.next()
    expect(delay).toBeCalledWith(1000)

    timeNow += 1001
    time.nowExact.mockReturnValue(timeNow)
    generator.next()
    lastArgs = put.mock.calls[put.mock.calls.length - 1]
    expect(lastArgs).toMatchSnapshot()
    generator.next()
    expect(delay).toBeCalledWith(1000)

    put.mockImplementation(() => {
      throw { message: 'CANCELLED' }
    })
    cancelled.mockReturnValue(true)
    expect(generator.next().done).toBe(false)
    put.mockReset()
    generator.next(true)
    lastArgs = put.mock.calls[put.mock.calls.length - 1]
    expect(lastArgs).toMatchSnapshot()
  })
})
