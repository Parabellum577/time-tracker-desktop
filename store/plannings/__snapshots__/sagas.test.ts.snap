// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Store] Plannings Sagas [Failed] Should emit GetOpenPlannings 1`] = `
Object {
  "done": false,
  "value": Object {
    "payload": Object {
      "code": 100,
      "message": "SOME_ERROR",
    },
    "type": "plannings/GET_OPEN_PLANNINGS_FAILURE",
  },
}
`;

exports[`[Store] Plannings Sagas [Success] Should update spent time every seconds 1`] = `
Array [
  Object {
    "error": undefined,
    "meta": undefined,
    "payload": Object {
      "PlanningID": 2,
      "additionalTime": 0,
      "field": "SpentOnline",
    },
    "type": "plannings/TIME_TICK_IN_ACTIVE_PLANNING",
  },
]
`;

exports[`[Store] Plannings Sagas [Success] Should update spent time every seconds 2`] = `
Array [
  Object {
    "error": undefined,
    "meta": undefined,
    "payload": Object {
      "PlanningID": 2,
      "additionalTime": 1.001,
      "field": "SpentOnline",
    },
    "type": "plannings/TIME_TICK_IN_ACTIVE_PLANNING",
  },
]
`;

exports[`[Store] Plannings Sagas [Success] Should update spent time every seconds 3`] = `
Array [
  Object {
    "error": undefined,
    "meta": undefined,
    "payload": Object {
      "PlanningID": 2,
      "additionalTime": 0,
      "field": "SpentOnline",
    },
    "type": "plannings/TIME_TICK_IN_ACTIVE_PLANNING",
  },
]
`;
